service/dap/server_test.go

```diff
@@ -1562,6 +1562,43 @@ func TestGlobalScopeAndVariables(t *testing.T) {
 	})
 }
 
+// TestShadowedVariables executes to a breakpoint and checks the shadowed
+// variable is named correctly.
+func TestShadowedVariables(t *testing.T) {
+	runTest(t, "testshadow", func(client *daptest.Client, fixture protest.Fixture) {
+		runDebugSessionWithBPs(t, client, "launch",
+			// Launch
+			func() {
+				client.LaunchRequestWithArgs(map[string]interface{}{
+					"mode": "exec", "program": fixture.Path, "showGlobalVariables": true,
+				})
+			},
+			// Breakpoints are set within the program
+			fixture.Source, []int{},
+			[]onBreakpoint{{
+				// Stop at line 13
+				execute: func() {
+					client.StackTraceRequest(1, 0, 20)
+					stack := client.ExpectStackTraceResponse(t)
+					expectStackFrames(t, stack, "main.main", 13, 1000, 3, 3)
+
+					client.ScopesRequest(1000)
+					scopes := client.ExpectScopesResponse(t)
+					expectScope(t, scopes, 0, "Arguments", 1000)
+					expectScope(t, scopes, 1, "Locals", 1001)
+					expectScope(t, scopes, 2, "Globals (package main)", 1002)
+
+					client.VariablesRequest(1001)
+					locals := client.ExpectVariablesResponse(t)
+
+					expectVarExact(t, locals, 0, "(a)", "a", "0", !hasChildren)
```

Comment:
Let's add the result of getting their evaluateName and using that in an expression to the test.
For good name, you should be able to use validateEvaluateName helper.
For bad name, you could get an error, then switch scope (as if a user did that in the UI) and then try again.

Reply:
By switching scope, do you mean running the program until after if statement so only one of the a's is in scope?

Reply:
It doesn't look like a separate scope object is created for the code the curlies, so running until after the curly would be the only option. Not really necessary. Without any extra special magic for evaluate expressions, you could just add two evaluate requests for "a" and "(a)" to show that they yield the same results. Not critical, but might be useful since we had the whole offline discussion if having () around the name makes a difference in expressions.